database/ffldb/db.go

```diff
@@ -1331,6 +1350,99 @@ func (tx *transaction) FetchBlockHeaders(hashes []chainhash.Hash) ([][]byte, err
 	return headers, nil
 }
 
+// GetBlockCount counts all blocks stored in the database. This function has to
+// iterate a cursor over all blocks and thus is relatively slow.
+//
+// The interface contract guarantees at least the following errors will
+// be returned (other implementation-specific errors are possible):
+//   - ErrTxClosed if the transaction has already been closed
+//
+// This function is part of the database.Tx interface implementation.
+func (tx *transaction) GetBlockCount() (uint32, error) {
+	// Ensure transaction state is valid.
+	if err := tx.checkClosed(); err != nil {
+		return 0, err
+	}
+
+	count := uint32(len(tx.pendingBlockData))
+	cursor := tx.blockIdx2Bucket.Cursor()
+	for ok := cursor.First(); ok; ok = cursor.Next() {
+		count++
+	}
+	return count, nil
+}
+
+// ForEachBlockHeader iterates through each stored block header and invokes
+// the passed function with the raw serialized bytes for each one.  The raw
+// bytes are in the format returned by Serialize on a wire.BlockHeader.  The
+// function is called with block headers ordered ascending by height from
+// the genesis block, so it is guaranteed to be called with a parent block
+// before any child blocks.
+//
+// WARNING: It is not safe to mutate data while iterating with this
+// method.  Doing so may cause the underlying cursor to be invalidated
+// and return unexpected keys and/or values.
+//
+// The interface contract guarantees at least the following errors will
+// be returned (other implementation-specific errors are possible):
+//   - ErrTxClosed if the transaction has already been closed
+//
+// NOTE: The slices returned by this function are only valid during a
+// transaction.  Attempting to access them after a transaction has ended
+// results in undefined behavior.  Additionally, the slices must NOT
+// be modified by the caller.  These constraints prevent additional data
+// copies and allows support for memory-mapped database implementations.
+//
+// This function is part of the database.Tx interface implementation.
+func (tx *transaction) ForEachBlockHeader(fn func(headerBytes []byte) error) error {
```

Comment:
Everything here looks sound, but seeing this complexity makes me wonder if the code (in a future PR) should be updated to store the partial block rows at the time `StoreBlock` is called along with the header, some dummy values for the on-disk location, and a flag to indicate if the block data is available, and then have `writePendingAndCommit` update the record accordingly once the block is written to disk and the location is known.

Since it would be in the underlying treaps at that point, a simple `.ForEach` could be used in here.  I think it's also going to have to do something similar to that anyways since I can foresee the need for a `StoreHeader` with full blown headers-first sync.